test/Test/Database/LSMTree/Internal/RawPage.hs

```diff
@@ -314,3 +318,175 @@ prop_single_entry (Ref.PageContentSingle k op) =
+      "∀ key page. maybe (maximum (rawPageKeys page) < key) (rawPageFindKey page key)" $
+        maybe
+        (property $ rawPageKeys rawpage `maximumIsLessThan` key)
+        (const $ property Discard)
+        (rawPageFindKey rawpage key)
+  where
+    maximumIsLessThan iVec obj
+      | null iVec = True -- When there are no elements, the test passes
+      | otherwise = maximum iVec < obj
+
+findKeyIndexCounterExample :: Testable prop => SerialisedKey -> RawPage -> String -> prop -> Property
+findKeyIndexCounterExample key rawpage lawStr = counterexample msg
+  where
+    pKeys = rawPageKeys rawpage
+    msg = case rawPageFindKey rawpage key of
+      Nothing -> unlines
+        [ "No key found"
+        , show $ pKeys
+        , if null pKeys
+          then "<NONE>"
+          else show $ maximum pKeys
+        , show key
+        , if null pKeys
+          then "<NONE>"
+          else  show $ maximum pKeys < key
+        ]
+      Just loc -> concat
+        [ "Relational law violated:\n  "
+        , lawStr
+        , "\n"
+        , unwords
+            [ "At entry№"
+            , show loc
+            , "found next {"
+            , show key
+            , "} > {"
+            , show $ rawPageIndex rawpage loc
+            , "} of page index "
+            , show loc
+            ]
+        , show $ pKeys
+        , if null pKeys
+          then "<NONE>"
+          else show $ maximum pKeys
+        , show key
+        , if null pKeys
+          then "<NONE>"
+          else  show $ maximum pKeys < key
+        ]
+
+-- |
+-- Helper data-type for generating arbitrary pages and offset keys
+-- with a useful distribution of hit and misses for testing purposes.
+data RawPageAndOffset = RawPageAndOffset RawPage SerialisedKey
+  deriving stock (Eq, Show)
+
+instance Arbitrary RawPageAndOffset where
+
+  arbitrary = do
+    Ref.PageContentOrdered kops <- arbitrary
+    let rawpage = fst . Ref.toRawPage $ Ref.PageContentFits kops
+    RawPageAndOffset rawpage <$> case kops of
+      -- If there are no kOps, then generate any offset key
+      [] -> Ref.toSerialisedKey <$> arbitrary
+
+      -- Otherwise, if at least one kOp exists then generate the offset key according to
+      -- the following probability distribution which is conditional on kOps:
+      --   * (1/2): Offset Key exists within the page
+      --   * (1/6): Offset Key goes before the page
+      --   * (1/6): Offset Key goes after the page
+      --   * (1/6): Offset Key missing within the page
+      --            (Assuming that there are two or more keys)
+      xs ->
+        let keys = fst <$> xs
+            keySmall = Ref.toSerialisedKey $ minimum keys
+            keyLarge = Ref.toSerialisedKey $ maximum keys
+
+            missingWithinRange
+              | keySmall == keyLarge = []
+              | otherwise =
+                let missingKey = do
+                      let n = length keys - 1
+                      afterIndex <- (`mod` n) <$> arbitrary
```

Comment:
Seems like a job for [chooseInt](https://www.stackage.org/haddock/lts-22.8/QuickCheck-2.14.3/Test-QuickCheck.html#v:chooseInt).